src/pu-emmc.c

```diff
@@ -881,5 +896,55 @@ pu_emmc_new(const gchar *device_path,
     if (!pu_emmc_parse_clean(self, root, error))
         return NULL;
 
+    /* Check if raw binary overwrites partitions */
+    if (self->partitions && self->raw) {
+        PuEmmcPartition *part = self->partitions->data;
+        gsize part_start = part->offset * self->device->sector_size;
+        gsize raw_start;
+        gsize raw_end;
+        struct stat st;
+        for (GList *b = self->raw; b != NULL; b = b->next) {
+            PuEmmcBinary *bin = b->data;
+            g_autofree gchar *path = NULL;
+
+            path = pu_path_from_uri(bin->input->uri, prefix, error);
+            if (path == NULL) {
+                g_prefix_error(error, "Failed parsing input URI for partition: ");
+                return NULL;
+            }
+            if (stat(path, &st) == -1) {
+                g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                            "Could not get file size of '%s'", bin->input->uri);
+                return NULL;
+            }
+            bin->input->size = st.st_size;
+            raw_start = bin->output_offset * self->device->sector_size;
+            raw_end = bin->input->size;
+            raw_end += raw_start;
+            raw_end -= bin->input_offset * self->device->sector_size;
+            if (raw_end > part_start) {
+                g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                                  "Raw binary overlaps with first partition");
+                return NULL;
+            }
+            /* Check against all binaries before the current one */
+            for (GList *i = b->prev; i != NULL; i = i->prev) {
+                PuEmmcBinary *bin2 = i->data;
+                gsize raw2_start = bin2->output_offset * self->device->sector_size;
+                gsize raw2_end;
+                raw2_end = bin2->input->size;
+                raw2_end += raw2_start;
+                raw2_end -= bin2->input_offset * self->device->sector_size;
+                if ((raw_start >= raw2_start && raw_end <= raw2_end) ||
+                    (raw_start <= raw2_start && raw_end > raw2_start) ||
+                    (raw_start < raw2_end && raw_end >= raw2_end)) {
+                    g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                                "Raw binary overlaps with other raw binary");
+                    return NULL;
+                }
+            }
+        }
+    }
+
```

Comment:
To make it more readable, I think this section should go into its own local function. In the future, we may want to add additional checks for e.g. the partitions and their input data, which does similar things. Then `pu_emmc_new()` should be kept clean and should not contain thorough implementations of checking as is.